src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,194 @@
+#include "ClpIrV1Decoder.hpp"
+
+#include <emscripten/emscripten.h>
+#include <emscripten/val.h>
+
+#include <clp/ffi/encoding_methods.hpp>
+#include <clp/ffi/ir_stream/decoding_methods.hpp>
+#include <clp/TimestampPattern.hpp>
+
+#include "spdlog/spdlog.h"
+#include "types.hpp"
+
+// Constants
+static constexpr char cIsNotFourByteEncErrorMessage[] = "Is not four byte encoding.";
+
+static constexpr size_t cDefaultNumCharsPerMessage = 512;
+static constexpr size_t cDefaultNumLogEvents = 500'000;
+static constexpr size_t cFullRangeEndIdx = 0;
+static constexpr size_t cLogLevelNone = 0;
+
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> ClpIrV1Decoder* {
```

Comment:
1. By further looking into [the Emscripten manual](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#raw-pointers), I believe the I should return specify `return_value_policy::take_ownership()` for the factory function in EMSCRIPTEN_BINDINGS, if we do want to keep the return type as a pointer. On the other hand, the `move` semantics seem also supported. I'll look more into `move` instead then.
2. It might not directly be supported although we can provide a wrapper in JS. Let's use exceptions for now as that what we do in our JS implementations (e.g. the already implemented JsonlDecoder.) I will take a look again when we enable exception details in a future PR.

Reply:
For 2 I get it. It's expected and I'm ok with not using boost outcome.
For 1, I think we should avoid using raw ptrs if possible.

Reply:
Gotcha. We now use smart pointer (unique_ptr) as the return type.